src/storage/engine/btree_engine.rs

```diff
@@ -0,0 +1,404 @@
+    }
+
+    pub fn default() -> Self {
+        let cfs = &[CF_WRITE, CF_DEFAULT, CF_LOCK];
+        Self::new(cfs)
+    }
+
+    pub fn get_cf(&self, cf: CfName) -> Arc<RwLockTree> {
+        let index = self.cf_names.iter().position(|&c| c == cf);
+        match index {
+            None => unreachable!(
+                "Not exist CF:[{}]! Please create it by BTreeEngine::new(&[CfName])!",
+                cf
+            ),
+            Some(i) => self.cf_contents[i].clone(),
+        }
+    }
+}
+
+impl Engine for BTreeEngine {
+    type Iter = BTreeEngineIterator;
+    type Snap = BTreeEngineSnapshot;
+
+    fn async_write(
+        &self,
+        _ctx: &Context,
+        modifies: Vec<Modify>,
+        cb: EngineCallback<()>,
+    ) -> EngineResult<()> {
+        if modifies.is_empty() {
+            return Err(EngineError::EmptyRequest);
+        }
+        cb((CbContext::new(), write_modifies(&self, modifies)));
+
+        Ok(())
+    }
+    fn async_snapshot(&self, _ctx: &Context, cb: EngineCallback<Self::Snap>) -> EngineResult<()> {
+        cb((CbContext::new(), Ok(BTreeEngineSnapshot::new(&self))));
+        Ok(())
+    }
+}
+
+impl Debug for BTreeEngine {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "BTreeEngine []",)
+    }
+}
+
+pub struct BTreeEngineIterator {
+    tree: Arc<RwLockTree>,
+    cur_key: Option<Key>,
+    cur_value: Option<Value>,
+    valid: bool,
+    bounds: (Bound<Key>, Bound<Key>),
+}
+
+impl BTreeEngineIterator {
+    pub fn new(tree: Arc<RwLockTree>, iter_opt: IterOption) -> BTreeEngineIterator {
+        let lower_bound = match iter_opt.lower_bound() {
+            None => Unbounded,
+            Some(key) => Included(Key::from_raw(key)),
+        };
+
+        let upper_bound = match iter_opt.upper_bound() {
+            None => Unbounded,
+            Some(key) => Excluded(Key::from_raw(key)),
+        };
+        let bounds = (lower_bound.clone(), upper_bound.clone());
+        Self {
+            tree: tree.clone(),
+            cur_key: None,
+            cur_value: None,
+            valid: false,
+            bounds,
+        }
+    }
+
+    fn seek_and_valid(&mut self, range: (Bound<Key>, Bound<Key>), next_back: bool) -> bool {
+        let tree = self.tree.read().unwrap();
+        let mut range = tree.range(range);
+        let item = if next_back {
+            range.next_back()
+        } else {
+            range.next()
+        };
+        match item {
+            Some((k, v)) if self.bounds.contains(k) => {
+                self.cur_key = Some(k.clone());
+                self.cur_value = Some(v.clone());
+                self.valid = true;
+            }
+            _ => {
+                self.valid = false;
+            }
+        }
+        self.valid()
+    }
+}
+
+impl Iterator for BTreeEngineIterator {
+    fn next(&mut self) -> bool {
+        let range = (Excluded(self.cur_key.clone().unwrap()), Unbounded);
+        self.seek_and_valid(range, false)
```

Comment:
In `tree.range()`, the tree will search twice for the `min_node` and ` max_node` if both bounds are specified, and it will skill if one bound is `Unbounded`.